### PR TITLE
Fix escape code regressions in template literals and close #136

### LINKED ISSUE #136: #129 not fixed correctly
Please reopen #129. 

All the failing test cases listed still do not work and there are a number of regressions in non-tagged template literals.

* https://github.com/terser-js/terser/issues/129#issuecomment-431609433
* https://github.com/terser-js/terser/issues/129#issuecomment-431611171

The [ES2018 revision of illegal escape sequences within tagged template literals](https://github.com/terser-js/terser/issues/129#issuecomment-430666089) has not been implemented.

### PR DESCRIPTION
**Pull request by fabiosantoscode**

### CODE REVIEW

**Comment by kzc** [comment]
I have another post with many new failing tests but github ate it. Will post as soon as github is stable. The parser still has several regressions in *non-tagged* template strings.

Could you please give this PR and its commits a more meaningful title like:
```
Fix escape code regressions in template literals
```
as the commit log is the only record of changes for the project.

**Comment by kzc** [comment]
Please apply these tests: https://github.com/terser-js/terser/issues/129#issuecomment-431836246

They correspond to the bug reports I listed in #129.

**Comment by kzc** [comment]
**Expected** Node 10 result:
```
$ echo 'console.log(`\xNOPE`);' | node-v10.4.1 
[stdin]:1
console.log(`\xNOPE`);
             ^^^^

SyntaxError: Invalid hexadecimal escape sequence
```

**Actual** incorrect Terser result:
```
$ echo 'console.log(`\xNOPE`);' | terser | node-v10.4.1 
NOPE

$ echo 'console.log(`\xNOPE`);' | terser
console.log(`\0NOPE`);
```

**Comment by fabiosantoscode** [comment]
I've applied your tests, and fixed the code until their result matched them. In the `\xNOPE` case and in several in the tests which I wrapped in TODO comments, it's impossible to know if the template string is tagged or not, to make the decision on whether to throw the exception or not. So I haven't implemented that. I think it's OK to assume users won't be using escape codes that don't work in nodeJS.

File: test/compress/template-string.js [hunk 1]
```diff
@@ -309,9 +309,9 @@ keep_raw_content_in_tagged_template: {
         evaluate: true
     }
     input: {
-        var foo = tag`\u0020\u{20}\u{00020}\x20\40\040 `;
+        var foo = tag`\u0020\u{20}\u{00020}\x20 `;
```
Comment (kzc) [inline test/compress/template-string.js]: Why was the octal escape removed?

The ES2018 template revision allows for invalid octal escapes within tagged template literals. This PR should not error out in such a case:

```
$ echo 'console.log(String.raw`\u0020\u{20}\u{00020}\x20\40\040`);' | node-v10.4.1
\u0020\u{20}\u{00020}\x20\40\040
```
```
$ echo 'console.log(String.raw`\u0020\u{20}\u{00020}\x20\40\040`);' | terser
Parse error at 0:1,22
console.log(String.raw`\u0020\u{20}\u{00020}\x20\40\040`);
                      ^
ERROR: Octal escape sequences are not allowed in template strings
```

File: test/compress/template-string.js [hunk 2]
```diff
@@ -558,3 +558,135 @@ baz`;
...
+}
+
+es2018_revision_of_template_escapes_1: {
+    options = {
+        defaults: true,
+    }
+    input: {
+        console.log(String.raw`\unicode \xerces`);
+    }
+    expect_exact: "console.log(String.raw\`\\unicode \\xerces\`);"
```
Comment (kzc) [inline test/compress/template-string.js]: Please use the following input as per [patch](https://github.com/terser-js/terser/issues/129#issuecomment-431836246):
```
        console.log(String.raw`\unicode \xerces \1234567890`);
```
It's valid ES2018 code.
```
$ echo 'console.log(String.raw`\unicode \xerces \1234567890`);' | node-v10.4.1
\unicode \xerces \1234567890
```

Comment (kzc) [reply test/compress/template-string.js]: Note that the invalid octal escaped is only permissible in ES2018 tagged template strings, not untagged template strings:
```
$ echo 'console.log(`\1234567890`);' | node-v10.4.1
[stdin]:1
console.log(`\1234567890`);
              ^^
SyntaxError: Octal escape sequences are not allowed in template strings.
```

**Comment by kzc** [comment]
Latest PR is getting closer, but still some issues with tagged template detection:

**Expected** in node 10:
```
$ echo 'var x={y:()=>String.raw}; console.log(x.y()`\4321\u\x`);' | node-v10.4.1
\4321\u\x
```
**Actual** (incorrect):
```
$ echo 'var x={y:()=>String.raw}; console.log(x.y()`\4321\u\x`);' | terser
Parse error at 0:1,43
var x={y:()=>String.raw}; console.log(x.y()`\4321\u\x`);
                                           ^
ERROR: Octal escape sequences are not allowed in template strings
```

**Comment by kzc** [comment]
Another issue found in latest PR:

**Expected** in node 10:
```
$ echo 'console.log(String.raw`\u{-1}`);' | node-v10.4.1
\u{-1}
```
**Actual** (incorrect):
```
$ echo 'console.log(String.raw`\u{-1}`);' | terser
Parse error at 0:1,22
console.log(String.raw`\u{-1}`);
                      ^
ERROR: Invalid hex-character pattern in string
```
Could you please add these 2 new tests?
```
untagged_template_with_ill_formed_unicode_escape: {
    input: `
        console.log(\`\\u{-1}\`)
    `
    expect_error: ({
        "name": "SyntaxError",
        "message": "Invalid hex-character pattern in string",
        "line": 2,
        "col": 20
    })
}

tagged_template_with_ill_formed_unicode_escape: {
    input: {
        console.log(String.raw`\u{-1}`);
    }
    expect_exact: "console.log(String.raw`\\u{-1}`)";
    expect_stdout: "\\u{-1}"
    node_version: ">=10"
}
```

File: test/compress/template-string.js [hunk 3]
```diff
@@ -686,3 +686,33 @@ invalid_hex_character_pattern: {
...
     })
 }
+
+invalid_unicode_patterns: {
+    input: `
+        "\\u{110000}"
+    `
+    expect_error: ({
+        "name": "SyntaxError",
+        "message": "Unicode reference out of bounce"
```
Comment (kzc) [inline test/compress/template-string.js]: s/bounce/bounds/

File: test/compress/template-string.js [hunk 4]
```diff
@@ -686,3 +686,33 @@ invalid_hex_character_pattern: {
...
+    })
+}
+
+invalid_unicode_patterns_2: {
+    input: `
+        "\\u{100000061}"
+    `
+    expect_error: ({
+        "name": "SyntaxError",
+        "message": "Unicode reference out of bounce"
```
Comment (kzc) [inline test/compress/template-string.js]: s/bounce/bounds/

File: test/compress/template-string.js [hunk 5]
```diff
@@ -686,3 +686,33 @@ invalid_hex_character_pattern: {
...
+    })
+}
+
+invalid_unicode_patterns_3: {
+    input: `
+        "\\u{fffffffffff}"
+    `
+    expect_error: ({
+        "name": "SyntaxError",
+        "message": "Unicode reference out of bounce"
```
Comment (kzc) [inline test/compress/template-string.js]: s/bounce/bounds/

Comment (kzc) [reply test/compress/template-string.js]:
```
$ grep bounce lib/*.js
lib/parse.js:                    parse_error("Unicode reference out of bounce");
```

Comment (fabiosantoscode) [reply test/compress/template-string.js]: I thought it was funny... But fixed it :)

File: test/run-tests.js [hunk 6]
```diff
@@ -314,7 +325,9 @@ function run_compress_tests() {
                         });
                         return false;
                     }
-                    if (test.reminify && !reminify(test.options, input_code, input_formatted, test.expect_stdout)) {
+                    if (test.reminify
+                        && semver.satisfies(process.version, ">0.12")
+                        && !reminify(test.options, input_code, input_formatted, test.expect_stdout)) {
```
Comment (kzc) [inline test/run-tests.js]: Please don't do this. reminify should work on all node versions.

**Comment by fabiosantoscode** [comment]
[this test](https://github.com/terser-js/terser/pull/137#issuecomment-432445443) is tricky to implement. I didn't foresee that one :/

**Comment by fabiosantoscode** [comment]
I think I'll make this error only pop up in the parsing phase, not tokeniser. The tokeniser doesn't know enough about context to tell me whether it's a tagged template string or not.

**Comment by kzc** [comment]
You can see the two case of the tagged TemplateLiteral in the ES2018 grammar:

https://www.ecma-international.org/ecma-262/9.0/index.html#sec-left-hand-side-expressions

```
MemberExpression[Yield, Await]:
    PrimaryExpression[?Yield, ?Await]
    MemberExpression[?Yield, ?Await] [Expression[+In, ?Yield, ?Await]]
    MemberExpression[?Yield, ?Await].IdentifierName
    MemberExpression[?Yield, ?Await] TemplateLiteral[?Yield, ?Await, +Tagged]
    SuperProperty[?Yield, ?Await]
    MetaProperty
    newMemberExpression[?Yield, ?Await] Arguments[?Yield, ?Await]
```
```
CallExpression[Yield, Await]:
    CoverCallExpressionAndAsyncArrowHead[?Yield, ?Await]
    SuperCall[?Yield, ?Await]
    CallExpression[?Yield, ?Await] Arguments[?Yield, ?Await]
    CallExpression[?Yield, ?Await] [Expression[+In, ?Yield, ?Await]]
    CallExpression[?Yield, ?Await].IdentifierName
    CallExpression[?Yield, ?Await] TemplateLiteral[?Yield, ?Await, +Tagged]
```

**Comment by fabiosantoscode** [comment]
@kzc, your test is a useful test case for terser: When I input that, it manages to remove var x, and the object it contains. I propose you open a new issue for it.

**Comment by fabiosantoscode** [comment]
@kzc those are not all the cases. For example "(2+3)``"

**Comment by fabiosantoscode** [comment]
But I believe I'm done. Check out the code.

**Comment by kzc** [comment]
> For example "(2+3)``"

Good find. Please add this patch to this PR:

```patch
diff --git a/lib/output.js b/lib/output.js
index 4cd27be..dbc35dc 100644
--- a/lib/output.js
+++ b/lib/output.js
@@ -1105,7 +1105,15 @@ function OutputStream(options) {
     });
 
     DEFPRINT(AST_PrefixedTemplateString, function(self, output) {
+        var tag = self.prefix;
+        var parenthesize_tag = tag instanceof AST_Arrow
+            || tag instanceof AST_Binary
+            || tag instanceof AST_Conditional
+            || tag instanceof AST_Sequence
+            || tag instanceof AST_Unary;
+        if (parenthesize_tag) output.print("(");
         self.prefix.print(output);
+        if (parenthesize_tag) output.print(")");
         self.template_string.print(output);
     });
     DEFPRINT(AST_TemplateString, function(self, output) {
diff --git a/test/compress/template-string.js b/test/compress/template-string.js
index aeea946..ae169ac 100644
--- a/test/compress/template-string.js
+++ b/test/compress/template-string.js
@@ -1,3 +1,19 @@
+tagged_template_parens: {
+    input: {
+        (a) `0`;
+        (a => b) `1`;
+        (a = b) `2`;
+        (a + b) `3`;
+        (a ? b : c) `4`;
+        (a, b, c) `5`;
+        (~a) `6`;
+        (a.b) `7`;
+        (a["b"]) `8`;
+        (a()) `9`;
+    }
+    expect_exact: "a`0`;(a=>b)`1`;(a=b)`2`;(a+b)`3`;(a?b:c)`4`;(a,b,c)`5`;(~a)`6`;a.b`7`;a["b"]`8`;a()`9`;"
+}
+
 template_strings: {
     beautify = {
         quote_style: 3
```

In case you're wondering, using `PARENS()` would be much more cumbersome and less efficient than the simple approach used in the patch.

**Comment by kzc** [comment]
> your test is a useful test case for terser: When I input that, it manages to remove var x, and the object it contains

Please add this fix for https://github.com/terser-js/terser/pull/137#issuecomment-433530506 to this PR:

```patch
--- a/lib/transform.js
+++ b/lib/transform.js
@@ -270,2 +270,3 @@ TreeTransformer.prototype = new TreeWalker;
     _(AST_PrefixedTemplateString, function(self, tw) {
+        self.prefix = self.prefix.transform(tw);
         self.template_string = self.template_string.transform(tw);
```

as well as this accompanying test in `test/compress/template-string.js`:

```
tagged_call_with_invalid_escape_2: {
    options = {
        defaults: true,
        toplevel: true,
    }
    input: {
        var x = {
            y: () => String.raw
        };
        console.log(x.y()`\4321\u\x`);
        let z = () => String.raw;
        console.log(z()`\4321\u\x`);
    }
    expect: {
        var x_y = () => String.raw;
        console.log(x_y()`\4321\u\x`);
        console.log((() => String.raw)()`\4321\u\x`);
    }
    expect_stdout: [
        "\\4321\\u\\x",
        "\\4321\\u\\x",
    ]
    node_version: ">=10"
}
```

**Comment by fabiosantoscode** [comment]
Mocha tests should be running. I didn't touch them. Perhaps mochallel is bugging out.

**Comment by kzc** [comment]
> Mocha tests should be running. I didn't touch them. Perhaps mochallel is bugging out.

Please take a look. I don't see any mocha test results for node 0.10, 0.12...

https://travis-ci.org/terser-js/terser/builds/446874045

node 0.10: https://travis-ci.org/terser-js/terser/jobs/446874046
node 0.12: https://travis-ci.org/terser-js/terser/jobs/446874047

File: test/mocha.js [hunk 7]
```diff
@@ -17,9 +18,7 @@ fs.readdirSync(testDir).filter(function(file) {
 
 module.exports = function() {
     mocha.run(function(failures) {
-        if (failures) process.on("exit", function() {
-            process.exit(failures);
-        });
+        if (failures) process.exitCode = failures;
```
[the remark this answers is not in the document]
Comment (fabiosantoscode) [reply test/mocha.js]: Yes, it does the same as the code that was there before.

[the remark this answers is not in the document]
Comment (kzc) [reply test/mocha.js]: FWIW, process.exitCode is not present in node 0.10. It was added in v0.11.8.

https://nodejs.org/api/process.html#process_process_exitcode

[the remark this answers is not in the document]
Comment (kzc) [reply test/mocha.js]: > Yes, it does the same as the code that was there before.

Not the case. If you have the following code, the stderr message will be printed whether there are `failures` or not:

```
         if (failures) process.exitCode = failures; 
         console.error("*** executed after `process.exitCode = failures`");
```

**Comment by fabiosantoscode** [comment]
@kzc patch applied. I think we're done?

**Comment by fabiosantoscode** [comment]
I've just simplified it a lot

File: test/mocha.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (fabiosantoscode) [reply test/mocha.js]: What was there before was a `process.exit()` call just in the "exit" event. It's the same as setting the exitCode.

**Comment by fabiosantoscode** [comment]
Can we have a look at that in a future PR?

File: test/compress/template-string.js [hunk 8]
```diff
@@ -793,3 +793,12 @@ tagged_template_with_ill_formed_unicode_escape: {
     expect_stdout: "\\u{-1}"
     node_version: ">=10"
 }
+
+tagged_template_with_comment: {
+    input: {
+        console.log(String.raw/*foo*/`\u`);
```
Comment (kzc) [inline test/compress/template-string.js]: Nice. Can you test the other tagged template form as well?
```
console.log((() => String.raw)()/*foo*/`\x`);
```

File: test/mocha.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (fabiosantoscode) [reply test/mocha.js]: Done, reverted it.

**Comment by kzc** [comment]
Although it'd be nice to add [another scenario](https://github.com/terser-js/terser/pull/137#discussion_r228719517) to a test, it's not pressing. This PR can be merged as is.

Another issue can be opened as to why we are not seeing mocha tests on Travis for node [0.10](https://travis-ci.org/terser-js/terser/jobs/447203450#L2554) and [0.12](https://travis-ci.org/terser-js/terser/jobs/447203451#L2554). Intentionally introducing an error in a mocha test would be informative - it would show us whether the mocha tests were run at all on those node versions on Travis, or whether they were run and their results were not output.

**Comment by kzc** [comment]
Found another ES2018 tagged template bug:

**Expected** on node 10:
```
$ echo '"use strict"; console.log(String.raw`\u\x\567`);' | node-v10.4.1
\u\x\567
```

**Actual** (incorrect) PR result:
```
$ echo '"use strict"; console.log(String.raw`\u\x\567`);' | terser
Parse error at 0:1,36
"use strict"; console.log(String.raw`\u\x\567`);
                                    ^
ERROR: Legacy octal escape sequences are not allowed in strict mode
```
Another incorrect variation:
```
$ echo 'console.log(String.raw`\u\x\567`);' | terser --module
Parse error at 0:1,22
console.log(String.raw`\u\x\567`);
                      ^
ERROR: Legacy octal escape sequences are not allowed in strict mode
```
`--module` implies `"use strict"`.

**Comment by kzc** [comment]
Test cases for [bug](https://github.com/terser-js/terser/pull/137#issuecomment-433709208) above.

```patch
tagged_template_valid_legacy_octal: {
    input: {
        console.log(String.raw`\u\x\567`);
    }
    expect_exact: 'console.log(String.raw`\\u\\x\\567`);'
    expect_stdout: "\\u\\x\\567"
    node_version: ">=10"
}

tagged_template_valid_strict_legacy_octal: {
    input: {
        "use strict";
        console.log(String.raw`\u\x\567`);
    }
    expect_exact: '"use strict";console.log(String.raw`\\u\\x\\567`);'
    expect_stdout: "\\u\\x\\567"
    node_version: ">=10"
}

tagged_template_valid_module_legacy_octal: {
    options = {
        module: true,     // implies "use strict"
    }
    input: {
        console.log(String.raw`\u\x\567`);
    }
    expect_exact: 'console.log(String.raw`\\u\\x\\567`);'
    expect_stdout: "\\u\\x\\567"
    node_version: ">=10"
}
```

**Comment by kzc** [comment]
A problem with the test `tagged_template_valid_module_legacy_octal` above is that there are no `parse` options in `test/run-tests.js` in order to set `parse: { module: true }`. There cannot be a per-test `parse` option because the entire test file must be parsed with the same parse options. 

So `tagged_template_valid_strict_legacy_octal` can be kept as is, and `tagged_template_valid_module_legacy_octal` would have to be converted to a mocha test to be effective.

**Comment by kzc** [comment]
> I've sent a separate PR for that, check out #143

I don't follow. I don't see a new PR for the bug https://github.com/terser-js/terser/pull/137#issuecomment-433709208 with the presently failing test [tagged_template_valid_strict_legacy_octal](https://github.com/terser-js/terser/pull/137#issuecomment-433711415).

All these interrelated PRs and failing tests in master is confusing.

**Comment by fabiosantoscode** [comment]
We don't need to test tagged_template_valid_module_legacy_octal as we already have tests that prove the modules turn strict mode on.

**Comment by kzc** [comment]
@fabiosantoscode Looks good. Assuming all tests are green, please squash and merge this PR.

I bet you didn't think the ES2018 tagged template revision was as complicated as it proved to be. :-)

But it did expose some other bugs along the way.

**Pull request merged**